src/farcasterd/runtime.rs

```diff
@@ -1907,7 +708,316 @@ impl Runtime {
+        self.registered_services
+            .iter()
+            .filter(|s| matches!(s, ServiceId::Syncer(..)))
+            .count()
+    }
+
+    fn connection_has_swap_client(&self, peerd: &ServiceId) -> bool {
+        self.trade_state_machines
+            .iter()
+            .filter_map(|tsm| tsm.get_connection())
+            .any(|client_connection| client_connection == *peerd)
+    }
+
+    fn count_connections(&self) -> usize {
+        self.registered_services
+            .iter()
+            .filter(|s| matches!(s, ServiceId::Peer(..)))
+            .count()
+    }
+
+    fn get_open_connections(&self) -> Vec<NodeAddr> {
+        self.registered_services
+            .iter()
+            .filter_map(|s| {
+                if let ServiceId::Peer(n) = s {
+                    Some(*n)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    fn match_request_to_syncer_state_machine(
+        &mut self,
+        req: Request,
+        source: ServiceId,
+    ) -> Result<Option<SyncerStateMachine>, Error> {
+        match (req, source) {
+            (Request::SweepMoneroAddress(..), _) | (Request::SweepBitcoinAddress(..), _) => {
+                Ok(Some(SyncerStateMachine::Start))
+            }
+            (Request::SyncerEvent(SyncerEvent::SweepSuccess(SweepSuccess { id, .. })), _) => {
+                Ok(self.syncer_state_machines.remove(&id))
+            }
+            _ => Ok(None),
+        }
+    }
+
+    fn match_request_to_trade_state_machine(
+        &mut self,
+        req: Request,
+        source: ServiceId,
+    ) -> Result<Option<TradeStateMachine>, Error> {
+        match (req, source) {
+            (Request::RestoreCheckpoint(..), _) => Ok(Some(TradeStateMachine::StartRestore)),
+            (Request::MakeOffer(..), _) => Ok(Some(TradeStateMachine::StartMaker)),
+            (Request::TakeOffer(..), _) => Ok(Some(TradeStateMachine::StartTaker)),
+            (Request::Protocol(Msg::TakerCommit(request::TakeCommit { public_offer, .. })), _)
+            | (Request::RevokeOffer(public_offer), _) => Ok(self
+                .trade_state_machines
+                .iter()
+                .position(|tsm| {
+                    if let Some(tsm_public_offer) = tsm.open_offer() {
+                        tsm_public_offer == public_offer
+                    } else {
+                        false
+                    }
+                })
+                .map(|pos| self.trade_state_machines.remove(pos))),
+            (Request::LaunchSwap(LaunchSwap { public_offer, .. }), _) => Ok(self
+                .trade_state_machines
+                .iter()
+                .position(|tsm| {
+                    if let Some(tsm_public_offer) = tsm.consumed_offer() {
+                        tsm_public_offer == public_offer
+                    } else {
+                        false
+                    }
+                })
```

Comment:
this will forward the request to a swap matching `public_offer` regardless of the swap's state.
tsm.consumed_offer() would for instance also work for `TradeStateMachine::SwapdRunning(SwapdRunning { public_offer, .. })`.

"Luckily", only `attempt_transition_to_swapd_launched` has a matching handler for `Request::LaunchSwap`, but in general, I think it's open how much of the checking `match_request_to_trade_state_machine` should handle:

In principle, all of the validation could happen here already, but I think we're better off only having it handle the minimum (e.g. ensuring that `swap_id` or `public_offer` match, which should be enough to have a unique & compatible match), and then leaving the remaining checks for the state transition function implementations.

But this necessitates that their `event.request` matchers can't ever have catch-alls that return new state, only logging & `None` or the original state like [here](https://github.com/farcaster-project/farcaster-node/pull/693/files#diff-e5f6d9e096b173ed0a096a10871b452ffe63db2298afffdc1b61652bd5f051dfR497-R510).

Reply:
The function `match_request_to_trade_state_machine` will disappear once the ServiceBus segregation lands as well. As long as we don't have unified identifiers we'll be stuck with this matching though. The way I see it, as little validation as possible should be applied within this auxiliary matching function. Any fine-grained logic should happen in the state machine handlers. Once the identifiers are unified, we can add a trait to each request that returns an identifier of sorts and allows as to immediately select a state machine without any further unpacking of the request.